axum-macros/src/typed_path.rs

```diff
@@ -0,0 +1,279 @@
+                ::axum::extract::Path::from_request(req).await.map(|path| path.0)
+            }
+        }
+    })
+}
+
+fn simple_pluralize(count: usize, word: &str) -> String {
+    if count == 1 {
+        format!("{} {}", count, word)
+    } else {
+        format!("{} {}s", count, word)
+    }
+}
+
+fn expand_unit_fields(ident: &syn::Ident, path: LitStr) -> syn::Result<TokenStream> {
+    for segment in parse_path(&path) {
+        match segment {
+            Segment::Capture(_, span) => {
+                return Err(syn::Error::new(
+                    span,
+                    "Typed paths for unit structs cannot contain captures",
+                ));
+            }
+            Segment::Static(_) => {}
+        }
+    }
+
+    Ok(quote_spanned! {path.span()=>
+        #[automatically_derived]
+        impl ::axum_extra::routing::TypedPath for #ident {
+            const PATH: &'static str = #path;
+        }
+
+        #[automatically_derived]
+        impl ::std::fmt::Display for #ident {
+            fn fmt(&self, f: &mut ::std::fmt::Formatter<'_>) -> ::std::fmt::Result {
+                write!(f, #path)
+            }
+        }
+
+        #[::axum::async_trait]
+        #[automatically_derived]
+        impl<B> ::axum::extract::FromRequest<B> for #ident
+        where
+            B: Send,
+        {
+            type Rejection = ::axum::http::StatusCode;
+
+            async fn from_request(req: &mut ::axum::extract::RequestParts<B>) -> Result<Self, Self::Rejection> {
+                if req.uri().path() == <Self as ::axum_extra::routing::TypedPath>::PATH {
+                    Ok(Self)
+                } else {
+                    Err(::axum::http::StatusCode::NOT_FOUND)
+                }
+            }
+        }
+    })
+}
+
+fn format_str_from_path(segments: &[Segment]) -> String {
+    segments
+        .iter()
+        .map(|segment| match segment {
+            Segment::Capture(capture, _) => format!("{{{}}}", capture),
+            Segment::Static(segment) => segment.to_owned(),
+        })
+        .collect::<Vec<_>>()
+        .join("/")
+}
+
+fn captures_from_path(segments: &[Segment]) -> Vec<syn::Ident> {
+    segments
+        .iter()
+        .filter_map(|segment| match segment {
+            Segment::Capture(capture, span) => Some(format_ident!("{}", capture, span = *span)),
+            Segment::Static(_) => None,
+        })
+        .collect::<Vec<_>>()
+}
+
+fn parse_path(path: &LitStr) -> Vec<Segment> {
+    path.value()
+        .split('/')
+        .map(|segment| {
+            if let Some(capture) = segment.strip_prefix(':') {
+                Segment::Capture(capture.to_owned(), path.span())
+            } else {
+                Segment::Static(segment.to_owned())
+            }
+        })
+        .collect()
+}
+
+enum Segment {
+    Capture(String, Span),
+    Static(String),
+}
+
+#[test]
+fn ui() {
```

Comment:
I think it would be nice to ignore this test on non-stable, but still allow it to be run using `cargo test -- --ignored` / `cargo test -- --include-ignored`:

```suggestion
#[test]
#[rustversion::attr(not(stable), ignore)]
fn ui() {
```

Then you shouldn't need the rustversion attributes inside the function.

Reply:
I think that makes sense but for some reason if I do

```rust
#[test]
#[rustversion::attr(not(nightly), ignore)]
fn ui() {
    let t = trybuild::TestCases::new();
    t.compile_fail("tests/typed_path/fail/*.rs");
    t.pass("tests/typed_path/pass/*.rs");
}
```

and then `cargo +nightly t -p axum-macros typed_path` the tests still runs. Don't see an issue on rustversion about this so not sure whats up.

Reply:
Yeah, it needs to be `not(stable)` instead of `not(nightly)` 😄